pkg/kubelet/lifecycle/handlers.go

```diff
@@ -105,6 +106,30 @@ func resolvePort(portReference intstr.IntOrString, container *v1.Container) (int
 	return -1, fmt.Errorf("couldn't find port: %v in %v", portReference, container)
 }
 
+// formatURL formats a URL from args.
+func formatURL(scheme string, host string, port int, path string) *url.URL {
+	u, err := url.Parse(path)
```

Comment:
parsing, then constructing a URL directly with what was previously simply appended to the request URL is very likely to produce different results than the current code path. See https://play.golang.org/p/gRoW1MFj8Ga for several examples of path values that would produce different results. If we're going to change how we construct these, we need comprehensive unit tests around the inputs to be sure we're not regressing something

Reply:
That's fair, and I can attempt to change `formatURL` to be compatible with the old format.

That said, I think it's worth pointing out that I have an expectation of HTTP-based probers and HTTP-based lifecycle handlers to conform to the same underlying logic, and that's not currently the case while probers use a similar [`formatURL`](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/prober/prober.go#L175) but lifecycle handlers do not. It feels more like this regression is a bug fix.

Reply:
> It feels more like this regression is a bug fix.

I can appreciate the perspective, but someone whose probes stop working because of this will not thank us for the "fix"

Reply:
Definitely; sorry, I wasn't trying to argue that we should ignore the problem as a result. I can see two possible paths forward though:

1. "Fix" `formatURL` to prevent a regression by ensuring unit tests validate that it produces the same URLs that it did before.
2. Allow for the regression, call it a bug fix, and put it behind a feature gate so it's opt-in (I'm not familiar with how to do this so may need some help here).

#1 seems easier, but #2 feels like the more correct answer.

Reply:
@liggitt Hopefully the latest push is a good compromise here. It adjusts how the URL gets created, which alleviates the regression in all ways I could think of, except for situations where the user provides a leading `/` in their path (previously, they would end up with `//path/to/resource`, but now they get `/path/to/resource`). I believe this is the most correct way to continue without regressing and without having so much change in the formatting of the URL that it requires a feature gate.